motoko/encrypted-notes-dapp/src/encrypted_notes_rust/src/lib.rs

```diff
@@ -0,0 +1,591 @@
+use candid::CandidType;
+use ic_cdk::api::{caller as caller_api};
+use ic_cdk::export::{candid, Principal};
+use ic_cdk::storage;
+use ic_cdk_macros::*;
+use serde::{Deserialize, Serialize};
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+use std::mem;
+use std::collections::btree_map::Entry::{Vacant, Occupied};
+
+type PrincipalName = String;
+type PublicKey = String;
+type Ciphertext = String;
+type DeviceAlias = String;
+
+#[derive(Clone, CandidType, Deserialize)]
+pub enum GetCiphertextError {
+    // Ensure API types in [encrypted_notes_rust.did] are named exactly as specified below
+    #[serde(rename = "notFound")]
+    NotFound,
+    #[serde(rename = "notSynced")]
+    NotSynced,
+}
+
+#[derive(Clone, CandidType, Deserialize)]
+pub enum Result {
+    // Ensure API types in [encrypted_notes_rust.did] are named exactly as specified below
+    #[serde(rename = "err")]
+    Err(GetCiphertextError),
+    #[serde(rename = "ok")]
+    Ok(Ciphertext),
+}
+
+/// Deriving CandidType or implementing it is necessary for
+/// almost everything IC - if you want your structs to
+/// Save in stable storage or serialize in inputs/outputs
+/// You should derive CandidType, Serialize, Deserialize.
+#[derive(Clone, CandidType, Serialize, Deserialize)]
+pub struct EncryptedNote {
+    id: u128,
+    encrypted_text: String,
+}
+
+/// There can only be one Type in stable storage at a time.
+/// We use this struct to represent the full CanisterState
+/// So we can serialize it to stable storage.
+#[derive(Clone, CandidType, Serialize, Deserialize)]
+struct CanisterState {
+    counter: u128,
```

Comment:
```suggestion
    // number of notes across all users (to avoid running out of memory)
    counter: u128,
```

Reply:
Actually, we don't seem to use the counter, or am I missing something?

Reply:
During canister upgrades, the `counter` field of `CanisterState` is simply a stable representation of the value stored in `NEXT_NOTE`.

Reply:
Could you add a comment about that, please? (also in Motoko)